executor/executor.cc

```diff
@@ -1220,7 +1254,28 @@ void copyin(char* addr, uint64 val, uint64 size, uint64 bf, uint64 bf_off, uint6
 
 bool copyout(char* addr, uint64 size, uint64* res)
 {
-	bool ok = false;
+	
+#if GOOS_windows
+	std::atomic_bool ok = false;
```

Comment:
What is the rationale for this change?
It seems that the current code should at least compile for windows.

Reply:
The reason for this edit is due to the absence of **__atomic_store_n** function in MSVC.  To replicate the atomic functions using the <atomic> library.

Reply:
I see.
Duplicating large chunks of code to compensate for one missing API is not good.
Unfortunately we stuck on some older versions of gcc for some OSes, so we probably can't use std::atomic on all OSes.
We either need some portable wrappers for atomic APIs.
But I wonder if we could make NONFAILING itself return success/fail. We would use it in one other place.
For gcc it can achieved along the following lines:
```
#define NONFAILING(...)                                              \
	({                                                            \
	        bool ok = true; \
		__atomic_fetch_add(&skip_segv, 1, __ATOMIC_SEQ_CST); \
		if (_setjmp(segv_env) == 0) {                        \
			__VA_ARGS__;                                 \
		} else ok = false;                                  \
		__atomic_fetch_sub(&skip_segv, 1, __ATOMIC_SEQ_CST); \
		ok; \
	})
```

and for msvc something like this should work (I think):
```
#define NONFAILING(...)                          \
        ([&]() { __try { __VA_ARGS__; } __except (EXCEPTION_EXECUTE_HANDLER) { return false; } return true; }())
```

Then we could use this place to just do `return NONFAILING(...)`

Reply:
Still relevant.

Reply:
@dvyukov Hi, I'm @Freakston's GSoC mentor. Can you explain the purpose of the __atomic_store_n()? The only explanation I see is that it is to protect the ok variable from being optimized away in a way that is incompatbile with setjmp/longjmp. But that seems like a very weird way of doing that. The usual approach would be to simply make that variable volatile.

Reply:
Volatile accesses are not ordered with any other memory accesses. So this code:
```
	    switch (size) {
		    case 1:
			    *res = *(uint8*)addr;
			    break;
		    case 2:
			    *res = *(uint16*)addr;
			    break;
		    case 4:
			    *res = *(uint32*)addr;
			    break;
		    case 8:
			    *res = *(uint64*)addr;
			    break;
		    default:
			    fail("copyout: bad argument size %llu", size);
	    };
            __atomic_store_n(&ok, true, __ATOMIC_RELEASE);
```

if the atomic operation is replaced with a volatile, can be compiled as:
```
	    switch (size) {
		    case 1:
			    ok = true;
			    *res = *(uint8*)addr;
			    break;
		    case 2:
			    ok = true;
			    *res = *(uint16*)addr;
			    break;
		    case 4:
			    ok = true;
			    *res = *(uint32*)addr;
			    break;
		    case 8:
			    ok = true;
			    *res = *(uint64*)addr;
			    break;
		    default:
			    fail("copyout: bad argument size %llu", size);
	    };
```

Then we will get ok set to true even if the actual memory access traps.

Reply:
Sorry for the long silence. Ahh, reordering. I wouldn't have thought a volatile access was allowed to be reordered across a memory write, but seemingly that is not prohibited by the standard.
I think the change to the macro you suggested would be a good solution. I'll work with @Freakston to get this fixed.